### PR TITLE
Dev/saars/host builder example

### PR DESCRIPTION
**Pull request by xiaomi7732**
Add an example for HostBuilder. Address issue #136.
@severisv: Free free to jump on for the code review if you want.

### CODE REVIEW

File: examples/HostBuilder/Readme.MD [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  * We make up the configuration file named [hostsettings.json](./src/hostsettings.json) just for fun. This turns out a big fun later in these 2 senses:
+    * hostsettings.json is not copied by default by the build. This causes crashes when the settings file is set to be required. We updated the project file ([HostBuilderExample.csproj](./src/HostBuilderExample.csproj)) to copy `*.json` file to the output folder.
+    * The Application Insights SDK uses the hardcoded name of `appsettings.json` and `appsettings.{EnvironmentName}.json`. Putting application insights related settings in hostsettings.json will not just work. We end up having to add back an [appsettings.json](./src/appsettings.json) and put application insights settings there.
+* In `ConfigureServices()` method:
+  * Microsoft.AspNet.Hosting.IHostingEnvironment needs to be injected and is required by application insights while what we have, `hostContext.HostingEnvironment` is a Microsoft.Extensions.Hosting.IHostingEnvironment object. By testing, it doesn't implement both interfaces. We end up creating a wrapper extension method: [ToAspNetCoreHostingEnvironment()](./src/Extensions.cs).
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply examples/HostBuilder/Readme.MD]: Thanks for the ask. I had the same question and it is asked here: [#765](https://github.com/Microsoft/ApplicationInsights-aspnetcore/issues/765). We will see.

File: examples/HostBuilder/Readme.MD [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        public PrintHelloService(TelemetryClient telemetryClient)
+        {
+            this._client = telemetryClient;
+        }
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            _asyncTimer = new Timer(state =>
+            {
+                string message = "Hello Hosted Service!!!";
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply examples/HostBuilder/Readme.MD]: Thanks for the question. I reviewed the code. Do you mean check the cancellation token before creating the timer object or upon the timer is invoked?
Personally, in either case, I think:

- The creation of the timer is a very short operation, I don't think we need the cancellation token;
- Upon invoking, either track event or flush doesn't expect to be long-haul logic;

That I don't think it necessary to deal with the cancellation token, right? Or do I miss anything important?

[the remark this answers is not in the document]
Comment (karolz-ms) [reply examples/HostBuilder/Readme.MD]: Sorry, you are right. I got confused and for some reason thought the cancellation token applies to the service running task, but it only applies to the service *starting* task.

File: examples/HostBuilder/Readme.MD (hunk 1 above)
[the remark this answers is not in the document]
Comment (karolz-ms) [reply examples/HostBuilder/Readme.MD]: So Liudmila says using AppInsights.AspNetCore package like this is not supported. We should probably rework this example?

File: examples/HostBuilder/Readme.MD [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                configHost.SetBasePath(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location));
+                configHost.AddJsonFile("hostsettings.json", optional: false);
+                configHost.AddEnvironmentVariables();
+            }).ConfigureServices((hostContext, services) =>
+            {
+                // Enable console logging for ILogger
+                services.AddLogging(cfg => { cfg.AddConsole(); });
+                // Microsoft.AspNetCore.Hosting.IHostingEnvironment is required by Application Insights SDK
+                services.AddSingleton<WebHosting.IHostingEnvironment>(provider => hostContext.HostingEnvironment.ToAspNetCoreHostingEnvironment());
+                services.AddApplicationInsightsTelemetry();
```
Comment (lmolkova) [inline examples/HostBuilder/Readme.MD]: I think it's a way to go and certainly helps with the scenario.
But it also brings a lot of unwanted dependencies and extra stuff to the web host.
E.g. if a user chose to have generic Host, I don't think they will be happy to get a dependency on AspNetCore.
There are things that are not applicable to generic hosts: telemetry initializers, request tracking stuff, env variables, etc...

A cleaner approach would be to either disable certain things (e.g. you can turn off requests tracking) or build DI container from scratch

Comment (xiaomi7732) [reply examples/HostBuilder/Readme.MD]: 👍 I think building a DI container from the scratch sounds like a clearer approach. Let me spend some time looking into it.

File: examples/HostBuilder/Readme.MD (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply examples/HostBuilder/Readme.MD]: 👍 Great. So we are on the same page now :-)

File: examples/HostBuilder/Readme.MD (hunk 3 above)
Comment (xiaomi7732) [reply examples/HostBuilder/Readme.MD]: Update: Tracking the implementation by [#708](https://github.com/Microsoft/ApplicationInsights-aspnetcore/issues/708).

**Comment by xiaomi7732** [comment]
This shall be address by #185

**Pull request closed without merging**